DAQ: added 16bit event info (flags) to FRD format

smorovic opened the pull request:
#### PR description:
Goal of these changes is to propagate data/MC flag to FedRawDataInputSource and EventAuxiliary for raw samples converted from offline datasets. Information is added to the Raw data (FRD format) header, where version is shrunk to use 16-bits and other 16 bits reused for flags (2 used initially).  Version is bumped to v6 of the format.

Raw Output Module, which is used for conversions of offline datasets to FRD, betect data/MC flag in the event stores it in raw output (FRD version needs to be set to 6 or higher in that case). Check for the ancient FRD v1 has also now been removed. Compatibility with current real FRD data (v5 generated by the DAQ) is fully preserved.

#### PR validation:

Tested write + read chain with the generated data (fake BU); with the patch it is possible to generate new format, but also old one. Both modes have been tested. With v6 being default generated format, functional check is now part of the "RunBUFU" unit test.

smorovic committed:
added 16bit FRD event flags (used firstly for mc/data flagging).
reduced version from 32 to 16 bits. compatibility removed for frd v1
and adding v6. Input sources adjusted to read 16 bits from version,
but FRD v6 data is not compatible with older versions of CMSSW

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18552

- This PR adds an extra 40KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18552/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18552/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

fwyzard commented on EventFilter/Utilities/interface/FedRawDataInputSource.h:
```diff
@@ -132,7 +132,7 @@ class FedRawDataInputSource : public edm::RawInputSource {
   typedef std::pair<InputFile*, InputChunk*> ReaderInfo;
 
-  uint32 detectedFRDversion_ = 0;
+  uint16 detectedFRDversion_ = 0;
```
```suggestion
  uint16_t detectedFRDversion_ = 0;
```
?

fwyzard commented on EventFilter/Utilities/plugins/FRDStreamSource.h:
```diff
@@ -41,7 +41,7 @@ class FRDStreamSource : public edm::ProducerSourceFromFiles {
   const bool verifyChecksum_;
   const bool useL1EventID_;
-  unsigned int detectedFRDversion_ = 0;
+  uint16 detectedFRDversion_ = 0;
```
```suggestion
  uint16_t detectedFRDversion_ = 0;
```

fwyzard commented on IOPool/Streamer/interface/FRDEventMessage.h:
```diff
@@ -57,14 +69,22 @@
   uint32 version_;
   uint32 run_;
   uint32 lumi_;
-  uint32 eventLow_;
```
is it OK to change the "v5" specification here ?

smorovic replied:
Yes, I cross-checked with the EvB code and apparently v5 was not defined correctly here. event ID here is 32-bit, and padding value was removed. That struct is (fortunately) unused, so it is for reference only.

fwyzard commented on IOPool/Streamer/interface/FRDEventMessage.h:
```diff
@@ -121,11 +145,17 @@ class FRDEventMsgView {
 
+  //can only be MC with output module flag
+  bool isRealData() const {
+    return !(flags_ & FRDEVENT_MASK_FROMOUTPUTMODULE) || (flags_ & FRDEVENT_MASK_ISREALDATA);
```
why not simply
```suggestion
    return (flags_ & FRDEVENT_MASK_ISREALDATA);
```
?

smorovic replied:
I'm wary of 0 being used as MC=true in production.
If someone updates the version in EvB to use v6 or higher, makes updates in CMSSW but forgets about this, we can end up having data flagged as MC in HLT..(if all values are left 0).

If you think it should be simpler, we can invert isRealData and store "isMC" as a flag.

fwyzard replied:
Ah, wait, I may have confused the bits...

So with this PR
  - `flags = 0x00` means that the flags are not saved (like in the old format) so it's real data
  - `flags = 0x01` means that the flags are saved, and it's real data
  - `flags = 0x11` means that the flags are saved, and it's MC
  - `flags = 0x10` should not be possible

?

I was thinking we could keep the existing behaviour unchanged:
  - `flags = 0x00` means real data (either implicitly like in the old format, or explicitly for real data)
  - `flags = 0x01` means MC
  - `flags = 0x10` is not used
  - `flags = 0x11` is not used


If somebody reads a v5 file with an old release, they always get "real data".
If somebody reads a v6 file with an old release, they get an error.
If somebody reads a v5 file with a new release, they always get "real data".
If somebody reads a v6 file with a new release, they get "real data" or "MC" according to the flags.

Did I miss any case that could be a concern ?

fwyzard replied:
:+1:

smorovic replied:
I will use a bit more generic FRDEVENT_MASK_ISGENDATA flag name - because there is also use case for DaqFakeReader which isn't MC, but random/fake fed data.

fwyzard commented on IOPool/Streamer/src/FRDEventMessage.cc:
```diff
@@ -27,22 +29,25 @@ FRDEventMsgView::FRDEventMsgView(void* buf)
-    size_ += sizeof(uint32);
-    ++bufPtr;
+  if (version_<2) {
+    throw cms::Exception("FRDEventMsgView") << "FRD version is " << version_ << " is no longer supported";
```
```suggestion
    throw cms::Exception("FRDEventMsgView") << "FRD version " << version_ << " is no longer supported";
```

fwyzard commented on EventFilter/Utilities/plugins/RawEventOutputModuleForBU.h:
```diff
@@ -70,7 +70,8 @@ RawEventOutputModuleForBU<Consumer>::RawEventOutputModuleForBU(edm::ParameterSet
       token_(consumes<FEDRawDataCollection>(edm::InputTag(label_, instance_))),
       numEventsPerFile_(ps.getUntrackedParameter<unsigned int>("numEventsPerFile", 100)),
-      frdVersion_(ps.getUntrackedParameter<unsigned int>("frdVersion", 3)),
+      frdVersion_(ps.getUntrackedParameter<unsigned int>("frdVersion", 5)),
```
why do you pick `5` and not `6` by default ?

smorovic replied:
It could be 6 but then it doesn't work out of box with older CMSSW versions. If there is no concern about this, then let's make 6 default.

fwyzard replied:
If you mean that files produces with the new version of the output modules would not be readable by older CMSSW release, I don't think that's a concern.

The two use cases I know about are:
  - hilton tests, that re-build the files each time
  - hlt/gpu performance tests, for which we would like the new flag anyway

smorovic replied:
ok, I'll push 6 as default in the commit I'm preparing.

smorovic committed:
FRD format v6 flags changes (thanks for suggestions from @fwyzard):

* invert isRealData flag, use isGenData flag. Outputmodule flag no longer necessary
* removed fillFRDFlags option, sufficient is to go to the v5 format to avoid id
* frd v6 is default in the writer module
* fixed comments and making variable types consistent

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18564

- This PR adds an extra 24KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18564/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18564/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

smorovic committed:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18565

- This PR adds an extra 44KB to repository

cmsbuild commented:
A new Pull Request was created by  @smorovic (Srecko Morovic) for master.

It involves the following packages:

EventFilter/Utilities
IOPool/Streamer


@perrotta, @smuzaffar, @Dr15Jones, @makortel, @emeschi, @cmsbuild, @slava77, @jpata, @smorovic can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @makortel, @wddgit this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

smorovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented on IOPool/Streamer/src/FRDEventMessage.cc:
```diff
@@ -27,22 +29,25 @@ FRDEventMsgView::FRDEventMsgView(void* buf)
-    ++bufPtr;
+  if (version_ < 2) {
+    throw cms::Exception("FRDEventMsgView") << "FRD version " << version_ << " is no longer supported";
   }
```
How practical is this exception? I suppose it would be thrown only when reading version 1 data where the lowest 16 bits of the word are `0` or `1`. (in which case, should `version_ > 6` be made to throw as well? or throw `if (version_ >= 32)` that was used above to deduce the version being 1?)

smorovic replied:
We could improve it to throw when lower 16-bit value is not [2,6]. Forward compatibility for higher versions is not needed. In addition, for [2,5] we can require flag bits to all be set to 0.

smorovic replied:
What I suggested has been committed. If there are no more comments on this, I will resolve this one.

cmsbuild commented:
+1
Tested at: 4624e0f58daca026eacf1e1d4d885c246d4799c2
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0e5336/9516/summary.html
CMSSW: CMSSW_11_2_X_2020-09-22-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

makortel commented on IOPool/Streamer/interface/FRDEventMessage.h:
```diff
@@ -15,6 +15,18 @@
+ *
+ * Version 6 Format:
+ *   uint16 - format version number
+ *   uint16 - eventf flags
```
Just to make sure that I understood correctly. From `FRDEventMsgView` and `RawEventOutputModuleForBU ` I understand that the flags are in the higher 16-bit part and the version number in the lower 16-bit part of the first 32-bit word. But here the version number becomes before the flags. And that is correct because of the (little) endianness, i.e. reading the first 16 bits gives the version number, and the second 16 bits gives the flags.

By the way, is there a typo?
```suggestion
 *   uint16 - event flags
```

smorovic replied:
The intention is to have 16-bit version before the flag word in the file. I will fix the code to always order the two variables in this respect.

In practice, the handling of the format hasn't been endian safe.  For example, with little-endian first byte of the header will be end up having the the version number, in big-endian it will be at later position. However, there is currently no use, AFAIK, for this format on big-endian architectures (only Intel/AMD used with it in DAQ for the time being).

I'll correct also the typo you mentioned.

fwyzard replied:
Power and ARM builds are also little endian.

smorovic replied:
👍

smorovic replied:
For a bit more clarity in the code, I switched to using 16-bit type pointers to write or read version and flags. If there is no remaining issue, I'll mark this resolved.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0e5336/9516/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2540471
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2540442
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

smorovic commented on EventFilter/Utilities/plugins/FRDStreamSource.cc:
```diff
@@ -65,7 +65,8 @@ bool FRDStreamSource::setRunAndEventInfo(edm::EventID& id,
 
   if (detectedFRDversion_ == 0) {
     fin_.read((char*)&detectedFRDversion_, sizeof(uint32_t));
```
Agreed, this is a mistake. It can be a 16-bit read (and then another one for flags).

smorovic committed:
- improved check in FRDEventMessage view: allow versions [2-6]. For
  version <=5 require flags to be 0.
- dereferencing uint16* type pointers when writing or reading version and flags
- fixed memory overflow when reading version in FRD Source (adjusted read to
  16-bits at the time)
- fixed a typo

smorovic committed:
disable TCDS FED check (creates a lot of warnings) on non-real data events

cmsbuild commented:
The code-checks are being triggered in jenkins.

smorovic committed:
code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18577

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #31543 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @emeschi, @cmsbuild, @slava77, @jpata, @smorovic can you please check and sign again.

smorovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 1fd22f6a3794389298fe0732fb1b289b8f057642
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0e5336/9525/summary.html
CMSSW: CMSSW_11_2_X_2020-09-23-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0e5336/9525/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2539438
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2539415
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

makortel commented on EventFilter/Utilities/plugins/FRDStreamSource.cc:
```diff
@@ -64,8 +64,10 @@ bool FRDStreamSource::setRunAndEventInfo(edm::EventID& id,
-    assert(detectedFRDversion_ > 0 && detectedFRDversion_ <= 5);
+    fin_.read((char*)&detectedFRDversion_, sizeof(uint16_t));
+    fin_.read((char*)&flags_, sizeof(uint16_t));
+    detectedFRDversion_ = detectedFRDversion_ & 0xffff;
```
It seems to me this masking does not do anything.

smorovic replied:
Not with the latest change. Corrected it now.

smorovic committed:
Update FRDStreamSource.cc

Removing obsolete mask as detectedFRDversion_ is now unsigned 16-bit variable.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31543/18648

- This PR adds an extra 16KB to repository

cmsbuild commented:
Pull request #31543 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @emeschi, @cmsbuild, @slava77, @jpata, @smorovic can you please check and sign again.

smorovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: cfdbe4379cbf265eeb57d5e746fe040867148ae6
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0e5336/9607/summary.html
CMSSW: CMSSW_11_2_X_2020-09-27-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0e5336/9607/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2539438
 - DQMHistoTests: Total failures: 13
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2539403
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

smorovic commented:
+1

makortel commented:
+1

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/31543	cfdbe43 
- changes in this PR are in primarily DAQ parts of the EventFilter/Utilities, with limited (~no) impact on reco parts of the code

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1

fwyzard commented:
@smorovic can I backport this to CMSSW 11.1.x ?
For the MWGRs and thus for the hilton tests we are still using the 11.1.x release series !

smorovic commented:
@fwyzard yes, I see no obstacle to backporting it.

fwyzard commented:
should I include also #30457 in the 11.1.x backport ?

boudoul commented:
Hi @fwyzard , thanks for noticing those PRs. These developments can be a good target for the November MWGR .

smorovic commented:
@fwyzard we can include also #30457
I am not sure if DataFormats changes are accepted outside of prereleases (if not, we can remove changes related to them which are minor).

smorovic commented:
@fwyzard  - there is also:
https://github.com/cms-sw/cmssw/pull/30206
But I think not needed in short term for 11_1_X.
